ardupilot_gz_bringup/launch/robots/iris.launch.py

```diff
@@ -148,10 +149,28 @@ def generate_launch_description():
         output="screen",
     )
 
+    # Transform
+    topic_tools_tf = Node(
+        package="topic_tools",
+        executable="transform",
+        arguments=[
+            "/gz/tf",
+            "/tf",
+            "tf2_msgs/msg/TFMessage",
+            "tf2_msgs.msg.TFMessage(transforms=[x for x in m.transforms if x.header.frame_id == 'odom'])",
+            "--import",
+            "tf2_msgs",
+            "geometry_msgs",
+        ],
+        output="screen",
+        respawn=True,
+    )
```

Comment:
Very nice way to solve this issue, we keep the tf tree ROS REP 105 compliant and are able to use gazebo's odometry plugin.

With this changes gazebo's pose plugin tree isn't being used at all, right ? And using the odometry plugin without the prefix makes it incompatible with the pose tree. Should we keep publishing that tree at all ?

Reply:
> With this changes gazebo's pose plugin tree isn't being used at all, right ?

Let me check that. I had thought we were using both: the pose plugin for the joint states and the odometry plugin for the `odom -> base_link` TF. I'll see what happens if I switch one off.

I've managed to get a build of `cartographer` and `cartographer_ros` for macOS arm64

- https://github.com/srmainwaring/cartographer/tree/wips/wip-1.0.0-macos-arm64
- https://github.com/srmainwaring/cartographer_ros/tree/wips/wip-ros2-macos-arm

and am working through the [`ardupilot_ros`](https://github.com/ArduPilot/ardupilot_ros) example.

There are some conflicts with this setup when the odometry is working again. I seem to be getting a couple of sources of the `odom -> base_link` map, or something similar, that causes the frame to flicker between two orientations. I'm looking into that now, but think we may want launch args that allow us to switch rviz and the odom map on/off depending on the use.

Reply:
Having launch args for that sounds very nice! 
With this working it would be possible to use Gazebo as the source for `odom->base_link` transform instead of `cartographer` in the [`nav2` example](https://github.com/ArduPilot/ardupilot_ros/pull/11)

Reply:
The joint states plugin is separate from the pose plugin, since robot_pose_publisher is already using the joint states topic it should be possible to simply remove the pose plugin without affecting the rest of the system.

Reply:
> The joint states plugin is separate from the pose plugin, since robot_pose_publisher is already using the joint states topic it should be possible to simply remove the pose plugin without affecting the rest of the system.

You're correct. We can drop the `gz::sim::systems::PosePublisher` but still need to retain the `ros_gz_bridge` mapping for

```yaml
- ros_topic_name: "gz/tf"
  gz_topic_name: "/model/iris/pose"
  ros_type_name: "tf2_msgs/msg/TFMessage"
  gz_type_name: "gz.msgs.Pose_V"
  direction: GZ_TO_ROS
```

which then only contains the `odom -> base_link` TF, so we could just use a `topic_tools` relay to pub from `/gz/tf` to `/tf`, and this would also eliminate the empty TF warning in the MAVProxy console.

Reply:
@pedro-fuoco I'm not having much luck getting the initial map orientation aligned to the actual pose.

I noticed that there is a ros2 fork of `cartographer` and  `cartographer_ros` which is the version used in the Humble release - so I've ported these to macOS for consistency. They also have the updated service interface with  the `initial_pose` option (see below).

The `use_odometry` option in `cartographer.lua` does not seem to correct the initial pose, even after adding a `remappings` entry to the `cartographer_node` in  `cartographer.launch.py`.

There is a facility to use the services to set an initial pose using the cartographer services:

```bash
# Finish trajectory 0
ros2 service call /finish_trajectory cartographer_ros_msgs/srv/FinishTrajectory "{trajectory_id: 0}"

# Start trajectory 1 with a rotation of (roll=0, pitch=0 , yaw=90) deg
ros2 service call /start_trajectory cartographer_ros_msgs/srv/StartTrajectory "{configuration_directory: '$(ros2 pkg prefix ardupilot_ros)/share/ardupilot_ros/config', configuration_basename: 'cartographer.lua', use_initial_pose: true, initial_pose: {orientation: {z: 0.70710678, w: 0.70710678}}}"
```

However, this also rotates the Odometry visual in rviz by 90 deg as well, so back to where we started.

A temporary solution is to align the copter initial pose with the x-axis, which results in a consistent `rviz` session. I can continue looking through cartographer issues etc. for info (setting the initial map pose is not documented), but I'd prefer to move on and complete a full review the nav2 work first.
 
Figure: `ardupilot_ros cartographer.launch.py` with consistent pose and map orientation.
![map-orientation](https://github.com/ArduPilot/ardupilot_gz/assets/24916364/870652b0-6a74-48af-9d04-daaff149d468)